### ISSUE TITLE
[Feature Request] Auto-Reconnect

### ISSUE THREAD

**Issue by particlebanana** [balderdashy/sails-mongo#118]
Migrated from https://github.com/balderdashy/waterline/issues/342

It appears that all the Waterline connections are formed on startup of Sails, and if MongoDB restarts throughout the lifetime of the Sails service, then the DB connections will all fail (i.e. they will not re-connect even if MongoDB comes back up).

This is a big issue because occasionally MongoDB may need to be restarted (e.g. due to changes in config).

Is there any way around this? We have written our own sails-mongo-multi adapter to allow our multi-tenant service to switch databases at run time based on each request. Perhaps I can add code to reconnect if I detect that the connection has been closed?

**Comment by Glavin001** [comment]
+1 Any more information would be greatly appreciated. Auto-reconnecting is important.

**Comment by leobudima** [comment]
+1, this is really important in any kind of production

**Comment by jstrika** [comment]
+1, We just had an outage due to  missing "auto_connect" feature. Is there any workaround?

**Comment by johnfking** [comment]
Not sure if this works as I have not tested it, but it seems the individual in the accepted answer has some kind of workaround.

http://stackoverflow.com/questions/24006188/replica-set-not-working-as-expected

**Comment by mkeremguc** [comment]
+1 for auto-reconnect

**Comment by SyntaxRules** [comment]
+1, working with a database that I have no control over reset times.

**Comment by devinivy** [comment]
We'll happily take a PR.  Any chance this is simply configurable with the mongodb driver?

**Comment by listepo** [comment]
In node-mongodb-native 2.0 
```
autoReconnect, {Boolean, default: true} Reconnect on error.
```
see http://mongodb.github.io/node-mongodb-native/2.0/reference/connecting/connection-settings/
and ```auto_reconnect``` - incorrect value(```autoReconnect``` - correct)
see https://github.com/balderdashy/sails-mongo/blob/master/lib/connection.js#L106

**Comment by devinivy** [comment]
Will take a PR :+1:

**Comment by carlosguillen** [comment]
+1 just discovered this problem in production

**Comment by winrid** [comment]
+1 discovered this issue as well

**Comment by devinivy** [comment]
Please try latest master.  We merged a fix in https://github.com/balderdashy/sails-mongo/pull/345.  @particlebanana release soon?

**Comment by lucj** [comment]
Will this also help to try several time to connect to the db if this last one if not up yet when sails is lifted ?

**Comment by sailsbot** [comment]
Thanks for posting, @particlebanana.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mkeremguc** [comment]
it doesnt reconnect for me:

`MongoError: topology was destroyed`

autoReconnect is true in socketOptions
sails-mongo@0.12.0

**Comment by luislobo** [comment]
Fixes reconnect issues with MongoDB, defaults now to 30 seconds.
It also adds reconnectTries, so you can configure both reconnectInterval and reconnectTries.

@mudrekh and I found the issue about MongoDB not reconnecting. Basically, sails-mongo/lib/adapter.js sets reconnectInterval=200 and by default reconnectTries is 30, so, it just tries for 6 seconds. If the delay is more than 6 seconds, it will not retry.

@particlebanana can you please check this? Thanks!

**Comment by luislobo** [comment]
https://github.com/balderdashy/sails-mongo/pull/416

**Comment by Mudrekh** [comment]
Interesting to note, if you set reconnectTries (the new property) to -1, the mongo driver will actually try to reconnect forever.

**Comment by particlebanana** [comment]
Published `0.12.1` which includes the patch from @luislobo.

**Comment by luislobo** [comment]
Great! Thanks for publishing it that quick!

**Comment by dandanknight** [comment]
Hi, I'm having reconnect issues also. I've updated to 0.12.1, but still doesn't want to keep trying to reconnect. Does the `reconnectTries` option need to be in the URL string, or alongside the url field, or in it's own config/options key?

Here's my sails config for connection...

````
 connections: {
        mongodbServer: {
            adapter : 'sails-mongo',
            url : "mongodb://dandanknight:N3wquay69@ds017514.mlab.com:17514/sails",
            reconnectTries : -1
        }
    }
`````

Also, is there any way to get the MongoDB events logged to my Sails log? Or do I have to create a separate DB connection and hook into those events?

Many thanks
